mart/attack/enforcer.py

```diff
@@ -97,17 +95,26 @@ def verify(self, input_adv, *, input, target):
 
 
 class Enforcer:
-    def __init__(self, constraints: dict[str, Constraint] | None = None) -> None:
-        self.constraints = constraints or {}
+    def __init__(self, **modality_constraints: dict[str, dict[str, Constraint]]) -> None:
+        self.modality_constraints = modality_constraints
+
+    def _enforce(self, input_adv, *, input, target, modality="constraints"):
+        # Set modality="constraints" by default, so that it is backward compatible with existing configs without modalities.
+        if isinstance(input_adv, torch.Tensor):
+            # Finally we can verify constraints on tensor, per its modality.
+            for constraint in self.modality_constraints[modality].values():
+                constraint(input_adv, input=input, target=target)
+        elif isinstance(input_adv, dict):
+            # The dict input has modalities specified in keys, passing them recursively.
+            for modality in input_adv:
+                self._enforce(
+                    input_adv[modality], input=input[modality], target=target, modality=modality
+                )
+        elif isinstance(input_adv, list) or isinstance(input_adv, tuple):
+            # The list or tuple input is a collection of sub-input.
+            for input_adv_i, input_i, target_i in zip(input_adv, input, target):
+                self._enforce(input_adv_i, input=input_i, target=target_i)
 
     @torch.no_grad()
-    def __call__(
-        self,
-        input_adv: torch.Tensor | tuple,
-        *,
-        input: torch.Tensor | tuple,
-        target: torch.Tensor | dict[str, Any] | tuple,
-        **kwargs,
-    ) -> None:
-        for constraint in self.constraints.values():
-            constraint(input_adv, input=input, target=target)
+    def __call__(self, input_adv, *, input, target, **kwargs):
+        self._enforce(input_adv, input=input, target=target)
```

Comment:
Why not just rename `_enforce` to `__call__`? Why the indirection?

Reply:
Okay I see you use recursion. But why not recurse using `self(...)`?

Reply:
Moved recursion to `self(...)`.

Reply:
A few tests are added.